components/raftstore/src/store/worker/snap_gen.rs

```diff
@@ -0,0 +1,285 @@
+        allow_multi_files_snapshot: bool,
+    ) {
+        fail_point!("before_region_gen_snap", |_| ());
+        SNAP_COUNTER.generate.start.inc();
+        if canceled.load(Ordering::Relaxed) {
+            info!("generate snap is canceled"; "region_id" => region_id);
+            SNAP_COUNTER.generate.abort.inc();
+            return;
+        }
+
+        let start = Instant::now();
+        let _io_type_guard = WithIoType::new(if for_balance {
+            IoType::LoadBalance
+        } else {
+            IoType::Replication
+        });
+
+        if let Err(e) = self.generate_snap(
+            region_id,
+            last_applied_term,
+            last_applied_state,
+            kv_snap,
+            notifier,
+            for_balance,
+            allow_multi_files_snapshot,
+        ) {
+            error!(%e; "failed to generate snap!!!"; "region_id" => region_id,);
+            SNAP_COUNTER.generate.fail.inc();
+            return;
+        }
+
+        SNAP_COUNTER.generate.success.inc();
+        SNAP_HISTOGRAM
+            .generate
+            .observe(start.saturating_elapsed_secs());
+    }
+}
+
+pub struct Runner<EK, R, T>
+where
+    EK: KvEngine,
+    T: PdClient + 'static,
+{
+    tiflash_stores: HashMap<u64, bool>,
+
+    engine: EK,
+    mgr: SnapManager,
+    router: R,
+    pd_client: Option<Arc<T>>,
+    pool: FuturePool,
+}
+
+impl<EK, R, T> Runner<EK, R, T>
+where
+    EK: KvEngine,
+    R: CasualRouter<EK>,
+    T: PdClient + 'static,
+{
+    pub fn new(
+        engine: EK,
+        mgr: SnapManager,
+        cfg: Arc<VersionTrack<Config>>,
+        router: R,
+        pd_client: Option<Arc<T>>,
+    ) -> Runner<EK, R, T> {
+        Runner {
+            tiflash_stores: HashMap::default(),
+            engine,
+            mgr,
+            router,
+            pd_client,
+            pool: YatpPoolBuilder::new(DefaultTicker::default())
+                .name_prefix("snap-generator")
+                .thread_count(
+                    1,
+                    cfg.value().snap_generator_pool_size,
+                    SNAP_GENERATOR_MAX_POOL_SIZE,
+                )
+                .build_future_pool(),
+        }
+    }
+
+    pub fn snap_generator_pool(&self) -> FuturePool {
+        self.pool.clone()
+    }
+}
+
+impl<EK, R, T> Runnable for Runner<EK, R, T>
+where
+    EK: KvEngine,
+    R: CasualRouter<EK> + Send + Clone + 'static,
+    T: PdClient,
+{
+    type Task = Task<EK::Snapshot>;
+
+    fn run(&mut self, task: Task<EK::Snapshot>) {
+        match task {
+            Task::Gen {
+                region_id,
+                last_applied_term,
+                last_applied_state,
+                kv_snap,
+                canceled,
+                notifier,
+                for_balance,
+                to_store_id,
+            } => {
+                // It is safe for now to handle generating and applying snapshot concurrently,
+                // but it may not when merge is implemented.
```

Comment:
Is this 7-year-old comment (from #1963) still relevant? I'm assuming no because merge has already been implemented. If there are constraints between snap gen and apply, it's not a good idea to move snap gen into a separate worker.

Reply:
As generation holds a snapshot, as far as I can see, it doesn't seem merge would break the safety. What do you think @overvenus

Reply:
> Is this 7-year-old comment (from https://github.com/tikv/tikv/pull/1963) still relevant?

No, it's an out dated comment.

Reply:
Sounds good, removed it